lib/rules.js

```diff
@@ -0,0 +1,280 @@
+    }
+
+    return false;
+  }
+
+  /**
+   * Evaluate all provided conditions.  Return true if the conditions are
+   * satisfied or false if any condition is not satisified
+   */
+  evaluate(satisfiers) {
+    if (this.conditions === null) {
+      return true;
+    }
+
+    if (typeof satisfiers !== 'object') {
+      this._throw(ErrorCodes.InvalidSatisfiers, 'satisfiers must be an object');
+    }
+
+    for (let condition of Object.keys(this.conditions)) {
+      if (!satisfiers[condition]) {
+        this._throw(ErrorCodes.InvalidSatisfiers, `required satisfier for condition ${condition} not specified`);
+      }
+      if (!this._evaluateCondition(this.conditions[condition], satisfiers[condition])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Return a list of satisfier keys which are required to evaluate this
+   * ruleset
+   */
+  requiredSatisfiers() {
+    return Object.keys(this.conditions);
+  }
+}
+
+/**
+ * Rule encapsulates how a rule is to be evaluated.
+ */
+class Rule {
+  constructor({ruleId, conditions, values, description}) {
+    if (typeof ruleId !== 'string') {
+      this._throw(ErrorCodes.InvalidRule, 'ruleId must be string');
+    }
+    this.ruleId = ruleId
+
+    if (typeof conditions !== 'object') {
+      this._throw(ErrorCodes.InvalidConditions, 'conditions must be an object');
+    }
+    this.conditions = new Conditions(conditions, this);
+
+    if (typeof values !== 'object') {
+      this._throw(ErrorCodes.InvalidValues, 'values must be an object');
+    }
+    this.values = values;
+
+    if (typeof description !== 'string') {
+      this._throw(ErrorCodes.InvalidRule, 'description must be a string');
+    }
+  }
+
+  _throw(code, msg) {
+    let err = new Error(msg);
+    err.code = code;
+    err.ruleId = this.ruleId || '<unknown>';
+    err.description = this.description || '<unknown>';
+    throw err;
+  }
+
+  /**
+   * Evaluate a rule's conditions against provided condition satisfiers and if
+   * the conditions are satisfied, set the rule's values to the provided target
+   * object
+   */
+  evaluate(satisfiers, target) {
+    if (this.conditions.evaluate(satisfiers)) {
+      assign(target, this.values);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Return a list of satisfier keys which are required to evaluate this
+   * ruleset
+   */
+  requiredSatisfiers() {
+    return this.conditions.requiredSatisfiers();
+  }
+}
+
+/**
+ * Assign into the target object, the properties from the values object.
+ * Assignments are nested, so if values has a property 'a.b.c: 1', the target
+ * will have a, a.b and a.b.c created as objects.  If property already exists
+ * on the target object, it is overwritten with the new property.  If a
+ * property was an object, a non-object can overwrite it and vice-versa.  If a
+ * values object property is the value `null`, that property on the target
+ * object is deleted
+ */
+function assign (target, values) {
```

Comment:
Is the difference between this and `Object.assign` that null values are deleted?

Reply:
Yes.  Object.assign is not a deep assignment, which we need.  It also doesn't handle null-deletions, so we need to have something which handles both.

There's a slight bug here around Array handling, which I've addressed in later commits.  Basically, arrays are copied as objects, so `['a', 'b']` becomes `{'0': 'a', '1': 'b'}`.